src/ctap/crypto_wrapper.rs

```diff
@@ -62,6 +94,185 @@ pub fn aes256_cbc_decrypt(
+    }
+
+    /// Returns the corresponding public key.
+    pub fn get_pub_key(&self) -> CoseKey {
+        match self {
+            PrivateKey::EcdsaKey(ecdsa_key) => CoseKey::from(ecdsa_key.genpk()),
+        }
+    }
+
+    /// Returns the encoded signature for a given message.
+    pub fn sign_and_encode(&self, message: &[u8]) -> Vec<u8> {
+        match self {
+            PrivateKey::EcdsaKey(ecdsa_key) => {
+                ecdsa_key.sign_rfc6979::<Sha256>(message).to_asn1_der()
+            }
+        }
+    }
+
+    /// The associated COSE signature algorithm identifier.
+    pub fn signature_algorithm(&self) -> SignatureAlgorithm {
+        match self {
+            PrivateKey::EcdsaKey(_) => SignatureAlgorithm::ES256,
+        }
+    }
+
+    /// Writes the key bytes.
+    pub fn to_bytes(&self) -> Vec<u8> {
+        match self {
+            PrivateKey::EcdsaKey(ecdsa_key) => {
+                let mut key_bytes = vec![0u8; 32];
+                ecdsa_key.to_bytes(array_mut_ref!(key_bytes, 0, 32));
+                key_bytes
+            }
+        }
+    }
+}
+
+impl From<PrivateKey> for cbor::Value {
+    fn from(private_key: PrivateKey) -> Self {
+        cbor_array![
+            cbor_int!(private_key.signature_algorithm() as i64),
+            cbor_bytes!(private_key.to_bytes()),
+        ]
+    }
+}
+
+impl TryFrom<cbor::Value> for PrivateKey {
+    type Error = Ctap2StatusCode;
+
+    fn try_from(cbor_value: cbor::Value) -> Result<Self, Ctap2StatusCode> {
+        let mut array = extract_array(cbor_value)?;
+        if array.len() != 2 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_INVALID_CBOR);
+        }
+        let key_bytes = extract_byte_string(array.pop().unwrap())?;
+        match SignatureAlgorithm::try_from(array.pop().unwrap())? {
+            SignatureAlgorithm::ES256 => PrivateKey::new_ecdsa_from_bytes(&key_bytes)
+                .ok_or(Ctap2StatusCode::CTAP2_ERR_INVALID_CBOR),
+            _ => Err(Ctap2StatusCode::CTAP2_ERR_INVALID_CBOR),
+        }
+    }
+}
+
+impl From<ecdsa::SecKey> for PrivateKey {
+    fn from(ecdsa_key: ecdsa::SecKey) -> Self {
+        PrivateKey::EcdsaKey(ecdsa_key)
+    }
+}
+
+/// Encrypts the given private key and relying part ID hash into a credential ID.
+///
+/// Other information, such as a user name, are not stored. Since encrypted credential IDs are
+/// stored server-side, this information is already available (unecrypted).
+///
+/// Also, by limiting ourselves to private key and RP ID hash, we are compatible with U2F for
+/// ECDSA private keys.
+pub fn encrypt_key_handle(
+    env: &mut impl Env,
+    private_key: &PrivateKey,
+    application: &[u8; 32],
+) -> Result<Vec<u8>, Ctap2StatusCode> {
+    match private_key {
+        PrivateKey::EcdsaKey(ecdsa_key) => {
+            let master_keys = storage::master_keys(env)?;
+            let aes_enc_key = crypto::aes256::EncryptionKey::new(&master_keys.encryption);
+            let mut plaintext = [0; 80];
```

Comment:
Is the format specified? Why a hole from byte 4 to 15?

Reply:
The format is not specified, I made it up. My train of thought was as follows:

1. We have to use multiples of 16 bytes for AES encryption.
2. Credential IDs should look like random data.
3. Everything should be encrypted.
4. The credential ID length must be a multiple of 16.
5. Since we need 1 new block anyway, maybe reserve it for future occasions like this, to make backwards compatibility easier.

Reply:
Sounds good, but then we should document the rationale for this format either here or where we do the opposite operation, given it's not standard.

Reply:
The encryption now explains its own version (moving the format description to its docstring), and the decryption explains what versions it can parse. FYI the legacy encryption still exists in code for tests, and likely all old versions will move there.